Get functional representation of Engine instance

coxjonc opened the pull request:
Added a `curry()` method to the DefaultEngine() class that returns a curried version of the `get_page_source()` method with all the state baked in. If the instance state is altered later, the function behavior will remain unchanged.

I also added a test to ensure that this behavior works as expected. To that end I also had to add a simple test server that just echoes the request headers. It adds a little overhead to running tests, and more importantly may interfere with CI unless we can get it to run automatically, but hopefully it will come in handy for testing some of the other crawlers as well.  

The other changes are small - added dependencies to requirements-dev.txt, and renamed setup() to setUp() because it wasn't being called automatically by unittest.

Fix #9

Commit:
Added curry() method to DefaultEngine, new tests

The curry() method allows the user to get the functional representation
of an Engine instance. The function will behave the same even if the
instance state changes later.

For testing purposes I also added a tiny flask server that
echoes the request headers.

Commit:
Added curry() method to DefaultEngine, new tests

The curry() method allows the user to get the functional representation
of an Engine instance. The function will behave the same even if the
instance state changes later.

For testing purposes I also added a tiny flask server that
echoes the request headers.

Commit:
Merge branch 'master' of github.com:geezhawk/rasp

Commit:
Removed unnecessary functools import

Commit:
Removed unnecessary flask import

Commit:
Fix accidental change to test URL

hank-bond commented on tests/test_engines.py:
```diff
@@ -12,10 +12,9 @@
 class TestDefaultEngine:
     @patch('rasp.base.DefaultEngine._session')
-    def setup(self, req_mock):
+    def setUp(self, req_mock):
```
We're using `py.test` for our test running suite, so the function names for these need not change.  Also assert statements don't work with `nose2`, as they won't be introspected properly.

coxjonc replied:
My bad I'll switch it back to setup - I saw nose2 in requirements and thought it was a typo

wdm0006 replied:
Yeah, we need to clean up the old nose2 stuff, ill open an issue

coxjonc replied:
fixed

hank-bond commented on requirements-dev.txt:
```diff
@@ -5,8 +5,9 @@
 
 # packages for testing and coverage
+flask
 nose2
```
This is my bad for not removing `nose2` from our dev reqs when we switched to `py.test` a while ago.

coxjonc replied:
np

hank-bond replied:
I also left it in the [CONTRIBUTING.md](https://github.com/anidata/rasp/blob/master/CONTRIBUTING.md) 😬

hank-bond commented on tests/tiny_server.py:
```diff
@@ -0,0 +1,12 @@
+import json
```
This is definitely a suboptimal way to unit test.  Why don't we just make a call to [this testing site](https://httpbin.org/headers) instead of rolling our own?  We can just record the response with `betamax` anyway.

wdm0006 replied:
That site is super convenient.  Probably can use it elsewhere in tests too.

coxjonc replied:
Yeah that site rules. Wish I'd known about it beforehand haha

coxjonc replied:
This issue has been cleared too since I removed the flask dependency and tiny_server.py

coxjonc replied:
Whoops hadn't actually committed that - but now flask server is gone :)

hank-bond commented on tests/test_engines.py:
```diff
@@ -33,17 +32,29 @@ def test_default_pull_source_valid_url(self):
+        self.engine.headers.update({'Content-Type': 'text/json'})
+        get_source = self.engine.curry()
+        self.engine.headers.update({'Content-Type': 'text/xml'})
+        page1 = get_source('http://127.0.0.1:5000/echo-headers/')
```
If we were really set on testing the state change with this `flask` method, we should at least wrap it with a `betamax` mocked session so we don't have to run the `flask` server every time.

hank-bond replied:
Also the name of `curry` should be changed to something a bit more descriptive that at least implies a relation to `get_page_source`.

coxjonc replied:
Nah let's go the httpbin route. We can use betamax there.

coxjonc replied:
agreed re method name

coxjonc replied:
OK check out the latest commit - removed flask and switched to using httpbin. I had to refactor the get_page_source() method a little because it wasn't adding the instance method's headers to the request unless headers were passed to the method. Still haven't taken care of the other issues but I'll get to them soon

hank-bond reviewed (changes requested):
Left some comments, and don't forget to rebase your changes on top of the code I just merged into `upstream/master`.

Also we're trying to keep the commit messages roughly to [this style guide](https://github.com/agis-/git-style-guide#messages).

P.S. I know you didn't request the review yet so apologies if this is overzealous.

Commit:
Switched to using httpbin

Commit:
Removed flask and fixed setup() method

hank-bond commented on rasp/base.py:
```diff
@@ -61,9 +75,10 @@ def get_page_source(self, url, params=None, headers=None):
+        temp = headers
+        headers = deepcopy(self.headers)
+        if isinstance(temp, dict):
             headers.update(temp)
```
I'm an idiot.  I wrote this code and I didn't realize I created a circular reference.

Look at it a little closer, the operation ends up being `deepcopy(self.headers).update(deepcopy(self.headers))`.  Instead, let's try something like this:
```python
merged_headers = deepcopy(self.headers)
if isinstance(headers, dict):
    merged_headers.update(headers)

response = self.session.get(
    url, params=params, headers=merged_headers
)
```

coxjonc replied:
Looks good to me.

hank-bond commented on README.md:
```diff
@@ -40,3 +40,9 @@ The 4 implemented engines are:
+To run the tests install the dev dependencies listed in requirements-dev.txt,
+start the test server by running 
+`python tests/tiny_server.py`, and run the test suite with
+`nose2`.
```
Don't forget to update this to reflect removing the actual `flask` code and the fact that we're using `py.test`.

hank-bond reviewed (changes requested):
Round 2

Commit:
Fixed README and script to set headers

coxjonc commented:
This is dumb but I'm kind of at a loss for what to name this method while making it specific without making it too long and unwieldy. Does `get_scraper_func()` sound reasonable?

Commit:
Gave curry method a more explicit name

Commit:
Merge remote-tracking branch 'upstream/master'

coxjonc commented:
Rebased on your changes

hank-bond commented:
> This is dumb but I'm kind of at a loss for what to name this method while making it specific without making it too long and unwieldy. Does get_scraper_func() sound reasonable?

Ya know, I don't have a great idea either and I think that will be fine. @wdm0006 might have a good idea. That said, I think we should make it a `staticmethod` and pass in self as the first argument.  I think that will make it a bit more clear what the intention of the function is.

wdm0006 commented:
`as_func`?

hank-bond commented:
I kind of like @wdm0006 suggestion of generalizing it.  We very well may want to use this funcitonality in future methods.

```python
eng = DefaultEngine(**kwargs)
func_get = DefaultEngine.as_func(eng, 'get_page_source')
```

Think about it, the functionality of what you're doing in this function doesn't have any inherent tie to `get_page_source`.

wdm0006 commented:
+1, thats a really good idea.

Commit:
Refactored get_func() as staticmethod

It makes more sense to treat get_func() as a method that isn't tied to a
particular instance of a DefaultEngine class. Instead, pass the class as
a parameter. Also, accept a method name as an argument so that the
method can be used to curry multiple methods, not just get_page_source()

Commit:
Rename self to cls

coxjonc commented:
Thanks for the comments guys. Per @baronvonbadguy I refactored `get_func()` as a static method in the latest commit, as well as giving it a parameter for a method name. Check it out and holler at me if there are any other changes that need to be made.

Commit:
Updated docstring

wdm0006 reviewed (commented):
Looks good to me 🚀

hank-bond commented on rasp/base.py:
```diff
@@ -45,6 +45,24 @@ def _session(self, *args, **kwargs):
+        method = None
+        class_name = TmpEngine.__class__.__name__
+        try:
+            method = getattr(TmpEngine, func_name)
```
You can just do 
```python
try:
    return getattr(TmpEngine, func_name)
```

coxjonc replied:
That's a good point! I'll update it. And sorry for playing fast and loose with the Docstring haha I'll get that fixed.

coxjonc replied:
See commit 4bd17b3

hank-bond reviewed (changes requested):
Last changes I promise.  We're using [Google style docstrings](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) for this project.  

I'll make sure to add that to CONTRIBUTING.md so folks know in the future.

Commit:
Google-style docstrings and refactored as_func()

hank-bond commented on rasp/base.py:
```diff
@@ -47,19 +47,25 @@ def _session(self, *args, **kwargs):
 
     @staticmethod
     def as_func(cls, func_name):
```
`func_name` should really be called `method_name`.

hank-bond commented on rasp/base.py:
```diff
@@ -47,19 +47,25 @@ def _session(self, *args, **kwargs):
     def as_func(cls, func_name):
-        """
-        Curries the "func_name" method
+        """Curries the specified function
```
Should be something like 
> Curries the specified method with the current state of the Engine.

coxjonc replied:
Fixed

hank-bond commented on rasp/base.py:
```diff
@@ -47,19 +47,25 @@ def _session(self, *args, **kwargs):
 
         Returns:
-            User-defined class method
+            User-defined class method, if successful
```
At this point it's returning a function 😉

coxjonc replied:
Good point!

hank-bond reviewed (commented):
1. These changes are optional.
2. Thank you for putting up with my pedantry.
3. I hate myself.

Commit:
Docstring improvements

Commit:
Func renamed to method

hank-bond reviewed (approved):
😌 💨

coxjonc commented:
hallelujah haha

coxjonc commented:
Go ahead and merge!
